### PR TITLE
MSET and MSETNX

### PR DESCRIPTION
**Pull request by ipapapa**
Adding support for multi-key operations like MSET and MSETNX.

### CODE REVIEW

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 1]
```diff
@@ -134,7 +134,7 @@ private MultiKeyOperation(final List<String> keys, final OpName o) {
             this.op = o;
         }
 
-        @Override
+		@Override
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: indentation. I just do not get what is the standard here.

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 2]
```diff
@@ -2705,15 +2708,53 @@ public String get(String s) {
...
+            return connPool.executeWithFailover(new MultiKeyOperation<Long>(Arrays.asList(keysvalues), OpName.MSETNX) {
+                @Override
+                public Long execute(Jedis client, ConnectionContext state) {
+                    return client.msetnx(keysvalues);
+                }
+            });
+        } else {
+            return connPool.executeWithFailover(new CompressionValueMultiKeyOperation<Long>(Arrays.asList(keysvalues), OpName.MSETNX) {
+                   @Override
+                   public Long execute(final Jedis client, final ConnectionContext state) throws DynoException {
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: Where are we compressing the values here?

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 3]
```diff
@@ -2705,15 +2708,53 @@ public String get(String s) {
...
+                @Override
+                public String execute(Jedis client, ConnectionContext state) {
+                    return client.mset(keysvalues);
+                }
+            });
+        } else {
+            return connPool.executeWithFailover(new CompressionValueMultiKeyOperation<String>(Arrays.asList(keysvalues), OpName.MSET) {
+                   @Override
+                   public String execute(final Jedis client, final ConnectionContext state) throws DynoException {
+                            return client.mset(keysvalues);
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: same, no compression happening here

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/MultiKeyCompressionOperation.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * @author ipapapa
+ *
+ * @param <CL>
+ * @param <R>
+ */
+public interface MultiKeyCompressionOperation<CL, R> extends Operation<CL, R> {
+
+	 String[] compressMultiKeyValue(ConnectionContext ctx, String... value);
+
+    String decompressValue(ConnectionContext ctx, String value);
```
Comment (shailesh33) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/MultiKeyCompressionOperation.java]: any reason why this doesnt take a list of strings?

Comment (ipapapa) [reply dyno-core/src/main/java/com/netflix/dyno/connectionpool/MultiKeyCompressionOperation.java]: In multi-key write operations, we receive `String... keyValues` which is a combination of keys and values and we return either a `String` or ` Long`. In multi-key read operations, the API receives a `String keys` which is only the keys that we want to be read and we return a List of values. The list of values is being transformed through the `CollectionUtils.transform` to each independent components (this is how you had set it up and it looks good to me). Then we take each value and we decompress. Hence, our `compressMultiKeyValue` has numerous inputs and has some logic to perform the compression like selecting every second element which is the values. The `decompressValue` does not have any special logic other than decompressing each value one-by-one

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 5]
```diff
@@ -241,48 +240,56 @@ public String decompressValue(String value, ConnectionContext ctx) {
...
+                          if ((2 * value.length()) > connPool.getConfiguration().getValueCompressionThreshold()) {
+                              newItems.add(i, ZipUtils.compressStringToBase64String(value));
+                              ctx.setMetadata("compression", true);
+                          }
+                          
+                      } catch (IOException e) {
+                          Logger.warn(
+                                  "UNABLE to compress [" + value + "] for key [" + getStringKey() + "]; sending value uncompressed");
+                      }
+                	 }
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: indentation

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 6]
```diff
@@ -241,48 +240,56 @@ public String decompressValue(String value, ConnectionContext ctx) {
...
-
-            return result;
+        public String[] compressMultiKeyValue(ConnectionContext ctx, String... keyValues) {
+        	 List<String> items = Arrays.asList(keyValues);
+           	 List<String> newItems = new ArrayList<String>();
+           	 
+              for (int i = 0 ; i < items.size() ; i++) {
+             	 if( i%2 == 0 ) {
+             		 String value = items.get(i);
+
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: please maintain that comment about 2

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 7]
```diff
@@ -241,48 +240,56 @@ public String decompressValue(String value, ConnectionContext ctx) {
...
      * <lh>String Operations</lh>
      * <li>{@link #mget(String...) MGET}</li>
+     * <li>{@link #mset(String...) MSET}</li>
+     * <li>{@link #msetnx(String...) MSETNX}</li>
      * </ul>
      *
      * @param <T>
      *            the parameterized type
      */
     private abstract class CompressionValueMultiKeyOperation<T> extends MultiKeyOperation<T>
```
Comment (shailesh33) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: may be rename this to alternate value Multi key operation etc? just an idea

**Comment by shailesh33** [comment]
LGTM

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java (hunk 5 above)
Comment (ipapapa) [reply dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: Fixed.

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java (hunk 6 above)
Comment (ipapapa) [reply dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: Done

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java (hunk 7 above)
Comment (ipapapa) [reply dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: It is not an alternate value. It is a key-value pair.

**Pull request merged**